src/tactic/fin_cases.lean

```diff
@@ -3,53 +3,127 @@ Copyright (c) 2018 Scott Morrison. All rights reserved.
+      (do
+        numeric ← option.is_some <$> try_core (unify ty `(ℕ) <|> unify ty `(ℤ) <|> unify ty `(ℚ)),
+        result ← cases_core e,
+        match result with
+        -- We have a goal with an equation `s`, and a second goal with a smaller `e : x ∈ _`.
+        | [(_, [s], _), (_, [e], _)] :=
+          do let sn := local_pp_name s,
+              ng ← num_goals,
+              -- tidy up the new value
+              tactic.interactive.conv (some sn) none
+                (to_rhs >> match with_list.nth 0 with
+                | (some h) := conv.interactive.change (to_pexpr h)
+                | _ := `[try { conv.interactive.simp ff [] [] }] >> when numeric `[try { conv.interactive.norm_num [] }]
+                end),
+              s ← get_local sn,
+              try `[subst %%s],
+              ng' ← num_goals,
+              when (ng = ng') (rotate_left 1),
+              fin_cases_at with_list.tail e
+        -- No cases; we're done.
+        | [] := skip
+        | _ := failed
+        end)-- <|> fail "Something went wrong while performing `cases`."
+    end)
+
+meta def fin_cases_at' (with_list : list expr) (e : expr) : tactic unit :=
+focus1 $ fin_cases_at with_list e
+
+meta def expr_list_to_list_expr : expr → tactic (list expr)
+| `(list.cons %%h %%t) := do t ← expr_list_to_list_expr t, return (h :: t)
+| `([]) := return []
+| _ := failed
 
 namespace interactive
 private meta def hyp := tk "*" *> return none <|> some <$> ident
+local postfix `?`:9001 := optional
 
 /--
-`fin_cases h` performs case analysis on a hypothesis `h : fin n`, where `n` is an explicit natural
-number. `fin_cases *` performs case analysis on all suitable hypotheses.
+`fin_cases h` performs case analysis on a hypothesis of the form
+`h : A`, where `[fintype A]` is available, or
+`h ∈ A`, where `A : finset X`, `A : multiset X` or `A : list X`.
+
+`fin_cases *` performs case analysis on all suitable hypotheses.
 
 As an example, in
 ```
 example (f : ℕ → Prop) (p : fin 3) (h0 : f 0) (h1 : f 1) (h2 : f 2) : f p.val :=
 begin
-  fin_cases p,
+  fin_cases *; simp,
   all_goals { assumption }
 end
 ```
-after `fin_cases p`, there are three goals, `f 0`, `f 1`, and `f 2`.
+after `fin_cases p; simp`, there are three goals, `f 0`, `f 1`, and `f 2`.
 -/
-meta def fin_cases : parse hyp → tactic unit
-| none := do ctx ← local_context,
-             ctx.mfirst fin_cases_at <|> fail "No explicit `fin n` hypotheses."
-| (some n) := do h ← get_local n, fin_cases_at h
+meta def fin_cases : parse hyp → parse (tk "with" *> texpr)? → tactic unit
+| none none := do ctx ← local_context,
+                  ctx.mfirst (fin_cases_at' []) <|> fail "No hypothesis of the forms `x ∈ A`, where `A : finset X`, `A : list X`, or `A : multiset X`, or `x : A`, with `[fintype A]`."
+| none (some _) := fail "Specify a single hypothesis when using a `with` argument."
+| (some n) with_list :=
+  do
+    h ← get_local n,
+    with_list ← match with_list with
```

Comment:
I can't try this out at the moment. But it looks like you're elaborating the `with` arguments alone, without regards to the expected type, right? This will fail if e.g. you have a list of `int`s and try to do `fin_cases using [1, 2, 3]`.

Reply:
Thanks for catching this. It's absolutely a problem; now fixed, with a test cases.